### PR TITLE
retry and forcefully kill update-ca-certificates on ubuntu

### PR DESCRIPTION
**Pull request by guoger**
[#111476046](https://www.pivotaltracker.com/story/show/111476046)

Signed-off-by: Zhou Xing <xingzhou@cn.ibm.com>

### CODE REVIEW

File: platform/cert/cert_manager.go [hunk 1]
```diff
@@ -40,9 +42,12 @@ type certManager struct {
 	updateCmdArgs []string
 	logger        logger.Logger
 	logTag        string
+	// Update execution time limit in seconds
+	// No retry if 0
+	updateTimeout int
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: lets change updateTimeout to time.Duration type.

File: platform/cert/cert_manager.go [hunk 2]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+			process, err := c.runner.RunComplexCommandAsync(command)
+			if err != nil {
+				return bosherr.WrapError(err, "Updating certificates")
+			}
+
+			resultChannel := process.Wait()
+
+			select {
+			case <-time.After(time.Duration(c.updateTimeout) * time.Second):
+				err = process.TerminateNicely(5 * time.Second)
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: do we have a test in exec_cmd_runner that makes sure that TerminateNicely does not error out when it's trying to kill process that's already gone?

File: platform/cert/cert_manager.go [hunk 3]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+			}
+		}
+
+		c.logger.Debug(c.logTag, "Failed to update new certificates")
+		return bosherr.WrapError(errors.New("Failed to update new certificates"), "Updating certificates")
+	} else {
+		c.logger.Debug(c.logTag, "Try to update new certificate files without retry")
+
+		_, _, _, err = c.runner.RunCommand(c.updateCmdPath, c.updateCmdArgs...)
+
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: extra newline

File: platform/cert/cert_manager.go [hunk 4]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+		if err != nil {
+			c.logger.Debug(c.logTag, "Failed to update new certificates")
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: remove log line. errors are already logged.

File: platform/cert/cert_manager.go [hunk 5]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+			return bosherr.WrapError(err, "Updating certificates")
+		}
+
+		c.logger.Debug(c.logTag, "Successfully updated new certificate files.")
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: remove log line. errors are already logged.

File: platform/cert/cert_manager.go [hunk 6]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+				}
+			case result := <-resultChannel:
+				if result.ExitStatus == 0 {
+					c.logger.Debug(c.logTag, "Successfully updated new certificate files.")
+					return nil
+				}
+			}
+		}
+
+		c.logger.Debug(c.logTag, "Failed to update new certificates")
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: remove log line. errors are already logged.

File: platform/cert/cert_manager.go [hunk 7]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+				return bosherr.WrapError(err, "Updating certificates")
+			}
+
+			resultChannel := process.Wait()
+
+			select {
+			case <-time.After(time.Duration(c.updateTimeout) * time.Second):
+				err = process.TerminateNicely(5 * time.Second)
+				if err != nil {
+					return bosherr.WrapError(err, "Updating certificates")
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: Change "Updating certificates" to "Killing update certificates cmd '%s'", updateCmdPath

File: platform/cert/cert_manager.go [hunk 8]
```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
...
+				}
+			case result := <-resultChannel:
+				if result.ExitStatus == 0 {
```
Comment (cppforlife) [inline platform/cert/cert_manager.go]: i think we should just check result.Error is nil for success.

File: platform/cert/cert_manager.go (hunk 2 above)
Comment (guoger) [inline platform/cert/cert_manager.go]: @cppforlife this one? https://github.com/cloudfoundry/bosh-utils/blob/develop/system/exec_cmd_runner_test.go#L378

**Pull request merged**